Masuda accounts

utugishinra opened the pull request:
アカウント登録画面を実装しました。

Commit:
registerクラス名修正

Commit:
アカウント登録画面：入力チェックエラー

Commit:
アカウント登録画面：フォームからユーザ名などの値を取得できるよう修正

Commit:
アカウント登録画面：エラーが無いときにエラーメッセージを出力しないように修正

Commit:
アカウント登録：登録時エラー

utugishinra committed:
アカウント登録画面作成完了

utugishinra committed:
アカウント登録：エラーメッセージを画面に表示する形式に修正

utugishinra committed:
テスト用の不要コードを削除

utugishinra committed:
Issue#11 アカウント登録テーブルのユニーク制約付与

utugishinra committed:
DBエラーメッセージが出力されなかったエラーに対応

KazukiYokoyama commented:
確認したところ、アカウント登録の機能について、幾つか満たされていないものがあります。

・アカウント登録に成功したら、アカウント登録完了に画面遷移する。
・既に登録されているアカウントの場合、「既に登録されているアカウントです」の文言を表示する。

上記を満たすよう修正をお願いします。
あともう少しだ。頑張って！

参考：
https://github.com/KazukiYokoyama/PHP_Study/wiki/%E3%82%A2%E3%82%AB%E3%82%A6%E3%83%B3%E3%83%88%E7%99%BB%E9%8C%B2

KazukiYokoyama commented on .vscode/launch.json:
```diff
@@ -0,0 +1,26 @@
+{
```
このファイルはローカルの開発環境だけで使用されるものです。
基本的に、各自の環境でのみ使う（更新される）ファイルはGitの追跡対象外にします。
.gitignoreファイルを追加し、.vscodeディレクトリを追跡しないようにしましょう。

参考：https://qiita.com/anqooqie/items/110957797b3d5280c44f

utugishinra replied:
了解しました。
gitの追跡リストから.vscodeを取り除きました。

KazukiYokoyama replied:
ＯＫです！

KazukiYokoyama commented on etc/Blog.sql:
```diff
@@ -26,13 +26,20 @@ FLUSH PRIVILEGES;
+CREATE TABLE `Accounts` (
+	`account_id` BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'アカウントID',
+	`account_name` VARCHAR(20) NULL DEFAULT NULL COMMENT 'アカウント名',
+	`email` VARCHAR(100) NULL DEFAULT NULL COMMENT 'メールアドレス',
```
account_nameとemailの属性にUNIQUEを追加するだけでも良いよ。
（やってることは同じ）

```suggestion
	`account_name` VARCHAR(20) NULL DEFAULT NULL UNIQUE COMMENT 'アカウント名',
	`email` VARCHAR(100) NULL DEFAULT NULL UNIQUE COMMENT 'メールアドレス',
```

utugishinra replied:
確かにそうですね。
ファイル内を見直して追加部分のみに修正しました。

KazukiYokoyama replied:
やったぜ！コードが短縮されると嬉しいね！

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+        // この中で登録処理を呼び出す
+        if (isset($_POST["insert"])){
+            // 入力チェック
+            $input_check = new Account();
```
コメントが不適切かもしれません。
実際には「Account」classのコンストラクタ内で入力チェックを行っています。
ここでは入力チェックの結果（エラーメッセージ等）を取得しています。

utugishinra replied:
修正しました。

KazukiYokoyama replied:
ごめん、書き方が悪かったね。
ここまで書いたんだから、この部分はこっちでコミットするっぺ。

```
           //  アカウント登録処理を開始
            $input_check = new Account();

           //  入力チェックの結果を取得する
            $this->page_data['username_errorMessage'] = $input_check->Get_account_name_errorMessage();
            $this->page_data['email_errorMessage'] = $input_check->Get_email_errorMessage();
            $this->page_data['password_errorMessage'] = $input_check->Get_password_errorMessage();

           //  入力チェックの結果エラーの判定
           if( !$this->page_data['username_errorMessage'] && 
               !$this->page_data['email_errorMessage'] && 
               !$this->page_data['password_errorMessage'])
            {
                      //  入力されたアカウントをDBに保存する処理…
            }
```

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+                $this->page_data['password_errorMessage'] = $password_message;
+            }else{
+                $input_check->Insert_Accounts();
+                $errormessage = $input_check->Get_errorMessage();
```
$input_check->Insert_Accounts();
↑これを意訳すると、
「入力チェック」インスタンスの「アカウント登録」メソッドを実行する。となります。
ちょっと分かりづらいので、インスタンス名を変更した方が良いかもしれません。

例えば、$account->InsertAccount();とすると、
「アカウント」インスタンスの「アカウント登録」メソッドを実行する。となります。

utugishinra replied:
了解しました。確かに分かりにくかったのでaccountに修正します。

KazukiYokoyama replied:
ＯＫ！確認しました。

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+
+}
+
+class Account {
```
classにもコメントをつけましょう。
特にこのclassは名前がシンプルな為、何を行うのか推察することが難しいです。

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+    private $account_name;              // ユーザ名
+    private $email;                     // メールアドレス
+    private $password;                  // パスワード
+    private $errorflg;                  // エラーフラグ
```
$errorflgちゃん、どこにも使われてないね…

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+        $this->Input_check();
+    }
+
+    function Insert_Accounts(){
```
コメントぉ…(´・ω・｀)

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+            if($check){
+                echo "<script>alert('アカウントの登録に成功しました！');</script>";
+            }else{
+                echo "<script>alert('アカウントの登録に失敗しました。');</script>";
```
すみません。純粋な疑問ですが、この処理が実行される契機ってどのようなものが考えられますか？
論理的にはSQL実行の結果、登録されたデータが無ければ、これが実行されると思うのですが…

単純にAccountsへのINSERTが失敗した場合、PDOExceptionのエラーでcatchの方へ流れてしまうので…

utugishinra replied:
すみません、コードの消し忘れです。消去しておきます。

utugishinra replied:
すみません、記述ミスです。修正しておきます。

KazukiYokoyama replied:
なるほど。承知しました。

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+                echo "<script>alert('アカウントの登録に失敗しました。');</script>";
+            }
+        }catch(PDOException $e){
+            $this->errorMessage .= 'データベース更新エラー\n'.addslashes($e->getMessage()).'。';
```
特に自作していないモジュールの例外をcatchする時は要注意です！

当然ですが、$eの中にはPDOException内で定義されたエラーメッセージが入っています。
しかし、そのエラーメッセージは、このサービスに対しては何の配慮もない内容です。
今回の場合、UNIQUEキー重複が発生すると、DBに実際に定義されているカラム名が表示されます。
エラーによってはDBの内容が類推可能となる内容が出力されてしまう可能性もあり、攻撃を受ける起点になる事もあります。

今回の場合は、画面には「アカウントの登録に失敗しました」と出力しておき、PDOExceptionからのエラーメッセージはphpのエラーログに出力するのが良いでしょう。

```suggestion
            error_log("[". date('Y-m-d H:i:s') . "]アカウント登録エラー：".addslashes($e->getMessage())."\n", 3, "/var/log/php/php_error.log");
            $this->errorMessage .= 'アカウントの登録に失敗しました。\nサイトの管理者に連絡してください。';
```

utugishinra replied:
了解しました、気を付けます。
今回はログ保存と表示を修正しておきます。

KazukiYokoyama replied:
ＯＫです。直ってるの確認しました。

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+        }
+    }
+
+    protected function Input_check() {
```
コメントぉ…(´・ω・｀)

protectedは、「このclassを継承した子classでも使用できます」という宣言です。
このclassがそもそも継承させる事を目的としない場合は、protected宣言は不要です。

Input_check()はこのclass内だけで使用しているみたいなので、
privateという宣言にすると、外部からは参照できなくなります。

```suggestion
    private function Input_check() {
```

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,127 @@
+    protected function Input_check() {
+        #ユーザ名：未入力チェック
+        if(empty($this->account_name)){
+            $this->account_name_errorMessage .= '<p>アカウント名を入力してください。</p>';
```
好みかと思うんだけど、エラーメッセージを記述する入力チェックに「～チェック」ってコメントは要らないと思うんだよね。エラーメッセージに書いてるし…

何の項目を入力チェックしているかっていうコメントがあれば良いと思います。

utugishinra replied:
コメントを修正しました。あと、細かいかもだけどコメント化に使っている記号を統一しときます。

KazukiYokoyama replied:
あれ、この修正コミットしてる？

KazukiYokoyama replied:
あがってきた！ＯＫです！

utugishinra committed:
不足機能の追加

・パスワードハッシュの形式をBcryptからDEFAULT形式に修正
・アカウント登録成功時にアカウント登録完了画面へ遷移するよう修正
・アカウント名、Eメールのユニークキー制約違反時に登録済みメッセージを表示

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,141 @@
+                $rtnFlg = 0;
+            }
+        }catch(PDOException $e){
+            $this->errorMessage = array(
```
私の感想としては、$errorMessageの内容が分かりづらいのではないかと感じます。
特に加工しないPDOExceptionの例外情報が欲しいのであれば、
$this->errorMessage = $eでも良いかと思います。

もう少し可読性を上げるための手段として、このようなものはどうでしょう？
```suggestion
            $this->errorMessage = '登録エラー';
            if($e->errorInfo[1] == 1062){
                        $this->errorMessage = '既に登録されているアカウントです！';
            }
```

utugishinra replied:
未加工の情報はログ出力する形式にしました。表示メッセージは上記の記述にしておきます。

KazukiYokoyama replied:
ＯＫです。確認しました。

KazukiYokoyama committed:
Update models/register.php

確かにprivateの方が役割として正しいですね。
private宣言に修正します。

Co-Authored-By: utugishinra <45139599+utugishinra@users.noreply.github.com>

utugishinra committed:
.vscodeを追跡対象から外した

utugishinra committed:
Blog.sqlを修正

utugishinra committed:
Merge branch 'masuda_accounts' of https://github.com/KazukiYokoyama/PHP_Study into masuda_accounts

utugishinra committed:
SQL実行時にエラーが出力されたため、修正

utugishinra committed:
入力チェックのコメントを修正した

入力チェック→入力チェックの結果取得に修正。
動作としてもこちらが正しい。

utugishinra committed:
誤ってetc/Blog.sqlを消してしまったため、再アップ

utugishinra committed:
レビュー結果反映

/
・ を  に修正
・不要な分岐：if()の消去
・catch(PDOException )のエラーメッセージ出力形式を変更

utugishinra committed:
アカウント登録画面修正

・エラーメッセージの取得方法を変更
・SQLエラー取得用の不要コードを削除

KazukiYokoyama committed:
コメントの修正とついでにコードの短縮を実施

KazukiYokoyama committed:
マージ作業

KazukiYokoyama commented on models/register.php:
```diff
@@ -2,13 +2,144 @@
+            $rtnFlg = 1;
+            
+        }catch(PDOException $e){
+            error_log("[". date('Y-m-d H:i:s') . "]アカウント登録エラー：".addslashes($e->getMessage())."\n", 3, "/var/log/php/php_error.log");
```
アカウントの重複による登録失敗は、業務処理の一環なので、エラーログには出さないようにしましょう。
後だしで申し訳ないが (´・ω・`) 一つそういう事で頼む。

原則としてエラーログには、プログラムのロジックエラーと、それを解析するために必要な情報を出力するのが良いと思います。

なので、アカウントの重複による登録失敗以外のエラーをログに書き出すように修正してください。

utugishinra replied:
ほほい、了解

KazukiYokoyama replied:
ＯＫやで！

utugishinra committed:
アカウント重複登録時、エラーログを残さないよう修正

utugishinra committed:
Merge branch 'masuda_accounts' of https://github.com/KazukiYokoyama/PHP_Study into masuda_accounts

KazukiYokoyama committed:
Merge branch 'master' into masuda_accounts
